qtbase: use -pthread instead of -lpthread

chuangzhu opened the pull request:
###### Description of changes

Change the linker option `-lpthread` to `-pthread`. This fixes ``undefined reference to `__atomic_*'`` errors when building `qt5.qtdeclarative` on `riscv64-linux`.

Related: https://bugs.gentoo.org/790689 https://bugs.gentoo.org/803470

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] riscv64-linux
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

SuperSandro2000 commented:
Please target staging

NickCao commented on pkgs/development/libraries/qt-5/modules/qtbase.nix:
```diff
@@ -116,6 +116,7 @@ stdenv.mkDerivation {
     sed -i '/PATHS.*NO_DEFAULT_PATH/ d' src/corelib/Qt5CoreMacros.cmake
     sed -i 's/NO_DEFAULT_PATH//' src/gui/Qt5GuiConfigExtras.cmake.in
     sed -i '/PATHS.*NO_DEFAULT_PATH/ d' mkspecs/features/data/cmake/Qt5BasicConfig.cmake.in
+    sed -i 's/-lpthread/-pthread/' mkspecs/common/linux.conf src/corelib/configure.json
```
Maybe add a link to upstream issue for future reference? Otherwise LGTM.

chuangzhu committed:
qtbase: use -pthread instead of -lpthread

This fixes "undefined reference to `__atomic_*'" errors when building
qt5.qtdeclarative on riscv64-linux.

SuperSandro2000 commented:
What does that change for x86-64_linux?

chuangzhu commented:
> What does that change for x86-64_linux?

It shouldn't affect other platforms. In the source code of GCC, `-pthread` [expands to `LD_AS_NEEDED_OPTION " -latomic " LD_NO_AS_NEEDED_OPTION` linker options on riscv](https://gcc.gnu.org/git/?p=gcc.git;a=blob;f=gcc/config/riscv/linux.h;h=38803723ba907103d8f6a093b98f0d769528259e;hb=HEAD#l40). On x86_64, `-pthread` only [expands to a `-D_REENTRANT` preprocessing option](https://gcc.gnu.org/git/?p=gcc.git;a=blob;f=gcc/config/i386/gnu-user-common.h;h=23b54c5be52510feb7d9383441b4073450b8b4aa;hb=HEAD#l39).

SuperSandro2000 commented:
I am not really confident merging this with the lack of qt and compiler knowledge I have.
